common/src/main/java/com/wynntils/mc/mixin/ChatComponentMixin.java

```diff
@@ -22,9 +41,151 @@ public abstract class ChatComponentMixin implements ChatComponentExtension {
     @Shadow
     private void refreshTrimmedMessages() {}
 
+    @Shadow
+    @Final
+    private List<GuiMessage.Line> trimmedMessages;
+
+    @Shadow
+    public abstract double getScale();
+
+    @Shadow
+    @Final
+    private Minecraft minecraft;
+
+    @Unique
+    private ChatTimestampFeature timestampFeature;
+
+    @Unique
+    private int timestampWidth = 0;
+
+    @WrapOperation(
+            method = "addMessageToDisplayQueue",
+            at = @At(value = "INVOKE", target = "Ljava/util/List;add(ILjava/lang/Object;)V"))
+    private void addMessageToDisplayQueue(
+            List<GuiMessage.Line> instance,
+            int i,
+            Object line,
+            Operation<Void> original,
+            @Local(name = "j") int index,
+            @Local(name = "message") GuiMessage message) {
+        GuiMessageExtension messageExtension = (GuiMessageExtension) (Object) message;
+
+        ((GuiMessageLineExtension) line).setCreated(messageExtension.getCreated());
+
+        WynntilsMod.postEvent(new AddGuiMessageLineEvent((GuiMessage.Line) line, index));
+
+        original.call(trimmedMessages, i, line);
+    }
+
     @Override
     public void deleteMessage(Component component) {
         allMessages.removeIf(guiMessage -> guiMessage.content().equals(component));
         refreshTrimmedMessages();
     }
+
+    @Unique
+    private ChatTimestampFeature getTimestampFeature() {
+        if (timestampFeature == null) {
+            timestampFeature = Managers.Feature.getFeatureInstance(ChatTimestampFeature.class);
+        }
+
+        return timestampFeature;
+    }
```

Comment:
The mixin looks good, but unfortunately accessing a manager instance like this shouldn't be done, especially from a mixin. What I'd propose is creating a new event, `RenderChatTimestampsEvent`, with it being disabled by default, and `ChatTimestampFeature` can listen to it and enable it. That easily solves the "reverse" dependency situation you have here.

Reply:
ok I've swapped it to use ChatTimestampFeature.RenderEvent.

Reply:
Just create a new event, and put it in the appropriate package in near the mixins package, inner class events are weird. As suggested above, `RenderChatTimestampsEvent` works, or any other describing name. One might ask why create an unique event for just this single use case, but this how the mod's architecture is.